File: algorithms/38.cpp
#include <iostream>
#include <sstream>
using namespace std;

class Solution
{
public:
    string countAndSay(int n)
    {
        string a("1");
        for ( int i = 1; i < n; i++ )
        {
            stringstream ss;
            char ch = a[0];
            int cnt = 1;
            for ( unsigned int j = 1; j < a.size(); j++ )
            {
                if ( a[j] == ch )
                    cnt++;
                else
                {
                    ss << cnt << ch;
                    ch = a[j];
                    cnt = 1;
                }
            }
            ss << cnt << ch;
            a = ss.str();
        }
        return a;
    }
};

int main()
{
    Solution solution;
    cout << solution.countAndSay(6) << endl;
    return 0;
}

File: algorithms/python/38.py
class Solution(object):
    def countAndSay(self, n):
        """
        :type n: int
        :rtype: str
        """
        def say(s):
            rls, idx = '', 0
            ch, cnt = s[0], 0
            while idx < len(s):
                if ch == s[idx]:
                    cnt += 1
                else:
                    rls += str(cnt) + ch
                    ch, cnt = s[idx], 1
                idx += 1
            rls += str(cnt) + ch
            return rls

        s = '1'
        for _ in range(n-1):
            s = say(s)
        return s
